src/main/java/org/apache/commons/compress/archivers/ArchiveStreamFactory.java

```diff
@@ -268,7 +278,15 @@ public static String detect(final InputStream in) throws ArchiveException {
             try {
                 tais = new TarArchiveInputStream(new ByteArrayInputStream(tarHeader));
                 // COMPRESS-191 - verify the header checksum
-                if (tais.getNextTarEntry().isCheckSumOK()) {
+                // COMPRESS-644 - do not allow zero byte file entries
+                TarArchiveEntry tae = tais.getNextTarEntry();
+                //try to find the first non-directory entry within the first 10 entries.
+                int count = 0;
+                while (tae != null && tae.isDirectory() && count++ < MAX_TAR_RECORDS) {
```

Comment:
This causes detection to fail for `src/test/resources/COMPRESS-612/test-times-star-folder.tar` (once you rebase on git master which tests more files).